### ISSUE TITLE
Stories with large images require scrolling to view captions

### ISSUE THREAD

**Issue by ghing** [denverfoundation/storybase#717]
This is captured from an e-mail thread about the viewer.

@bpawlak said:

> Our story viewer presents itself quite well save for the fact that the images aren't sized to 
> the viewport, resulting in required scrolling.  Is there a way to adjust that so that the 
> images (and captions) fit without scrolling?

Update: pending issues
- [x] Test with Fusion tables embed, YouTube, Vimeo, SoundCloud, small image/long caption
- [x] Strip dataset info from popup caption

**Comment by ghing** [comment]
via @jwirfs-brock:

> Good point, Bill. That's definitely been a concern. Another related concern is that there's no way for users to resize their images in the builder -- they have to resize them before uploading them (which pretty much no one does).

**Comment by ghing** [comment]
I'm assuming you're referring to image heights.  Image widths should be scaled with CSS to fit the various layouts.

It's always possible to dynamically resize image heights to fit the viewport height with JavaScript, but I think doing this would probably make for some undesired results in some cases.  

I'd be interested in looking at examples of stories where the image sizes/scrolling seems problematic, looking at what needs to be different with the images to fix this and then thinking what features/training is needed to support those changes.

**Comment by ghing** [comment]
Examples via @jwirfs-brock:
- [ELK Day of Service Project](http://floodlightproject.org/en/stories/elk-day-of-service-project/)
- [Nurturing Possibilities](http://floodlightproject.org/en/stories/nurturing-possibilities/)
- [Be a Friend, Make a Friend](http://floodlightproject.org/stories/be-a-friend-to-those-who-are-bullied-and-make-a-fr/)

**Comment by ghing** [comment]
Via @bpawlak:

> But in this example: http://floodlightproject.org/en/stories/elk-day-of-service-project/viewer/#sections/525b90d0d52d46a6a113cefd5b8001d1  it's presented in a full-screen window and the scrolling is still a bit of a pain.

**Comment by ghing** [comment]
This is what the image looks like in the example raised by @bpawlak:

![Screenshot from 2013-04-16 11:46:42](https://f.cloud.github.com/assets/110420/386978/46313766-a6b5-11e2-8a2c-e1a2e9e6bc70.png)

I did a quick image-edited mockup of what this might look like with the image scaled so it fits in the window:

![Screenshot from 2013-04-16 11:29:57](https://f.cloud.github.com/assets/110420/386980/58377344-a6b5-11e2-99d9-3b71024c5bf3.png)

Personally, I prefer the dynamic nature of the large image, and the way it looks, taking up the full width of the window, to the scaled, centered image.

What do you all think?

**Comment by ghing** [comment]
One easy fix might be to add a `title` attribute to the image and set it to the caption.  That way, when the user mouses over the big image, they will see the caption text:

![Screenshot from 2013-04-16 11:51:59](https://f.cloud.github.com/assets/110420/387005/29b6e51c-a6b6-11e2-9358-b16e1837b6bd.png)

**Comment by jwirfs-brock** [comment]
I like @ghing's suggestion about a caption/title text. This is similar to what happens when you are viewing a photoset on Flickr. Where would the best placement be? I think putting it in the center of the image covers up too much of the content -- maybe putting it at the top left would work.

**Comment by jwirfs-brock** [comment]
Example: http://www.flickr.com/search/?q=boulder+trails

**Comment by bpawlak** [comment]
Top left probably makes the most sense because of the scrolling issue (if we place it near bottom, you might not be able to see it.

**Comment by ghing** [comment]
@jwirfs-brock, is a tooltip ok (which pops up wherever the user first mouses over the image), or do you want something else that's styled to enhance visibility, and something that's consistently positioned in the same place. What should trigger the caption being shown? If you're not sure, that's fine and I can try to prototype a few different things.

**Comment by jwirfs-brock** [comment]
@ghing So, is the distinction that a tooltip would appear wherever the mouse is, whereas another solution (i.e., a caption that appears in the top left) would always be in the same place?

My initial reaction is that:
- the caption should appear in a consistent location, 
- and that location should be the top left. 
- It should appear whenever a user mouses over the image.

But, I might have a different opinion when I actually see it. I think the best step forward is to prototype this so we can take a look at it. Is that doable?

**Comment by ghing** [comment]
@jwirfs-brock,

> So, is the distinction that a tooltip would appear wherever the mouse is, whereas another solution (i.e., a caption 
> that appears in the top left) would always be in the same place?

Yeah, that's the main UI difference.  There are some implementation differences in terms of complexity, but my feeling is that a consistently positioned caption is helpful.  I'll work on a prototype in the next couple of days.  The first step I'm going to take is to look at existing captions to get a few test cases that cover the minimum, maximum and typical cases of caption length.

The only concern I have with the concept of always showing the caption on mouseover is that it could be annoying and also obscure the photo, but I'll reserve judgement until I've tried this out.

**Comment by jwirfs-brock** [comment]
@ghing Yeah, that's a good point about not wanting the caption to obscure the image itself. Let's just keep an eye on that and see if it's an issue once we have the prototype.

**Comment by ghing** [comment]
I wrote [a simple program](https://gist.github.com/ghing/0adb86e2477638ccce71) to find example captions that represent the min, max and average caption lengths and this is what I found as good samples for prototyping:

Min: 20 ('The Holly Fair (ULC)') in story /stories/the-holly-gets-a-face-lift/

Max: 429 ('In the past two decades, the number of births per year in the Children's Corridor increased nearly 50 percent. The proportion of births to mothers without a high school education also grew. In 1990, 26% of births to mothers over 20 were to mothers without a high school degree. Births to low-education mothers peaked in 2002, when 45% of mothers over 20 had not completed high school. Those children are now in elementary school.') in story /stories/the-health-education-gap-crosses-generations/

Avg: 68 ('Family Dollar on Colorado Blvd. in Northeast Park Hill (Jon Denzler)') in story /stories/the-pictures-of-the-food-desert/

**Comment by ghing** [comment]
Feedback on CSS only solution from @jwirfs-brock:
- Caption should  always be visible at bottom. This will prevent "jumping" around 
- Should test on very dark and very light images and a variety of images in general

**Comment by jwirfs-brock** [comment]
@ghing In data assets (maps, charts, tables) that have associated data, right now the entire caption (including the link to the associated data) is showing up in the pop-up. Here's an example:

http://dev.floodlightproject.org/en/stories/why-stories-matter-a-reflection-on-the-denver-ca-2/
(see section 5)

What are your thoughts on this? I'm not sure if it makes sense to include the links in the pop-up. It might make more sense just to include descriptive elements, and have the data links only at the bottom.

There are more examples in http://dev.floodlightproject.org/en/stories/anyone-can-be-a-hans-rosling/ on sections 4 and 5. For those, it definitely makes sense just to have the caption in the pop-up.

**Comment by jwirfs-brock** [comment]
@ghing For YouTube videos, having the pop-up caption is a bit odd.

Take a look at http://dev.floodlightproject.org/en/stories/anyone-can-be-a-hans-rosling/ on section 2. The YouTube caption appears at the exact same spot as our pop-up caption.

Also see http://dev.floodlightproject.org/en/stories/americorps-attendance/ on section 2.

I would recommend that for YouTube embeds, we drop the pop-up caption (if that's possible).

**Comment by jwirfs-brock** [comment]
@ghing I found an example of a small image with a long caption. It's in http://dev.floodlightproject.org/en/stories/americorps-attendance/ on section 10.

Here's a screen shot:
![screen shot 2013-10-31 at 11 34 36 am](https://f.cloud.github.com/assets/1020518/1447094/fd09857e-4241-11e3-870a-7126044c7c57.png)

Yes, it does look a little funny. But I didn't feel like I _had_ to mouse over the image. It's easy to mouse off if you do, anyway. So I think in this case it's fine. 

Basically, the small image, large caption case "passes."
